### PR TITLE
[Macros] Update plugin search options serialization

### PR DESCRIPTION
**Pull request by rintaro**
Previously plugin search options were serialized for each option kind. Instead serialize them in the order specified.

rdar://110903149

### CODE REVIEW

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

File: lib/Serialization/ModuleFileSharedCore.cpp [hunk 1]
```diff
@@ -126,18 +126,27 @@ static bool readOptionsBlock(llvm::BitstreamCursor &cursor,
...
-      break;
-    case options_block::COMPILER_PLUGIN_EXECUTABLE_PATH:
-      extendedInfo.addCompilerPluginExecutablePath(blobData);
+    case options_block::PLUGIN_SEARCH_OPTION: {
+      unsigned optKind;
+      StringRef optStr;
+      options_block::ResilienceStrategyLayout::readRecord(scratch, optKind);
+      switch (PluginSearchOptionKind(optKind)) {
+      case PluginSearchOptionKind::PluginPath:
+        optStr = "-plugin-path";
```
Comment (rintaro) [inline lib/Serialization/ModuleFileSharedCore.cpp:135]: Not sure about this. But I didn't wanted to introduce another "kind" enum just for `swift/Serialization/Validation.h`

Comment (DougGregor) [reply lib/Serialization/ModuleFileSharedCore.cpp:135]: The fact that LLDB needs to match on these strings feels a little brittle. Would we be better off exporting `PluginSearchOptionKind` into a header and having that in the list of plugin search options?

**Review by DougGregor — APPROVED** [review]

File: lib/Serialization/ModuleFormat.h [hunk 2]
```diff
@@ -651,6 +651,16 @@ enum class MacroIntroducedDeclNameKind : uint8_t {
...
 
+// These IDs must \em not be renumbered or reordered without incrementing
+// the module version.
+enum class PluginSearchOptionKind : uint8_t {
+  PluginPath,
+  ExternalPluginPath,
+  LoadPluginLibrary,
+  LoadPluginExecutable,
+};
+using PluginSearchOptionKindField = BCFixed<3>;
```
Comment (xymus) [inline lib/Serialization/ModuleFormat.h:662]: It looks like a `BCFixed<2>` should be enough for the current enum size.

**Review by xymus — APPROVED** [review]
The change looks good!

File: lib/Serialization/ModuleFileSharedCore.cpp (hunk 1 above)
Comment (rintaro) [reply lib/Serialization/ModuleFileSharedCore.cpp]: 👍 Added a commit.
Made `PluginSearchOption` a struct with `PluginSearchOption::Kind` in `SearchPathOptions.h`

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

**Review by bnbarham — APPROVED** [review]
LGTM other than @xymus comment

**Comment by rintaro** [comment]
`BCFixed<3>` is for possible future addition of another options without incrementing module format version.
Since this record is _not_ going to be used hundreds of times for each module, @xymus and I agreed we keep it 3 bits.

**Pull request merged**